### ISSUE TITLE
Excessive memory usage

### ISSUE THREAD

**Issue by SoFiA-Admin** [SoFiA-Admin/SoFiA#138]
When trying to process a 12.6 GB data cube with 60 GB of allocated memory, the process failed presumably after running out of memory. This occurred some time into SoFiA’s attempt to reload the data cube before parameterisation. The memory allocation corresponds to about 4.8 times the size of the data cube and should have been sufficient. SoFiA successfully managed to process a 6.3 GB subcube (half the size of the full cube).

Modules switched on during the run include noise scaling (local), S+C finder, merging, parameterisation and reliability calculation, with output of mask, moment maps and ASCII + XML catalogues.

Several things can be checked to determine the cause of the excessive memory usage:

* What is the actual memory usage of SoFiA in each step?
* Are intermediate data products no longer required actually being deleted?
* Can memory usage be reduced by manually invoking the garbage collector at critical steps?

**Comment by paoloserra** [comment]
If I remember correctly the cube reloading step was not part of SoFiA back when we measured the memory requirements for the SoFiA paper.

Regardless, the lines below in sofia_pipeline.py were our attempt to free up the memory taken by np_Cube, but my understanding is that they won't necessarily free up the memory right away.

	del np_Cube, dict_Header
	kwargs = Parameters["import"].copy()
	kwargs.update({"cubeOnly":True})
	np_Cube, dict_Header = import_data.read_data(Parameters["steps"]["doSubcube"], **kwargs)

I think these lines are the best candidate troublemakers.

**Comment by SoFiA-Admin** [comment]
Thanks, I will have a look at some of these lines to check how they each perform memory-wise. In addition, I have built a memory tracker into `sofia_pipeline.py` that can be enabled by setting `track_memory_usage = True` at the beginning of the file. The pipeline will then report its own peak memory usage at various points in time. I am currently running the memory tracker on the 13 GB cube to check how much SoFiA is using. I will report on the outcome once SoFiA terminates.

**Comment by SoFiA-Admin** [comment]
Here is the result of the first memory test on the full 12.6 GB cube. The plot shows the **peak** memory usage of SoFiA up to a given point in time (not the actual usage at that time).

![memory_profile_12 6gb](https://user-images.githubusercontent.com/6572706/40095568-4966638e-58bc-11e8-94b3-f836c420ffc6.png)

Much to my surprise, simply loading the cube creates a 36 GB memory spike, which is equivalent to almost 3 times the cube size! Noise scaling the adds just a little, while the S+C finder pushes the peak usage to almost 4 times the cube size. Merging and reliability calculation don’t add any more, but SoFiA then gets killed when trying to reload the data cube for parameterisation, as this will presumably push it above the 60 GB memory allocation limit.

In summary, SoFiA uses way too much memory, and I am particularly surprised about how much memory it takes to simply read in the cube at the beginning. I am running additional tests at the moment with slightly smaller sub-cubes to check at which point SoFiA breaks.

**Comment by paoloserra** [comment]
> I have built a memory tracker into sofia_pipeline.py that can be enabled by setting track_memory_usage = True at the beginning of the file.

That's very useful, thanks!

**Comment by SoFiA-Admin** [comment]
Well, things are starting to get a little crazy with SoFiA. What happened? Well, I managed to process the full cube in one go with 60 GB of allocated memory.

![memory_profile_12 6gb-subcube](https://user-images.githubusercontent.com/6572706/40097912-539c7fe4-58c8-11e8-8449-21236277919f.png)

How is this possible? Well, it turns out that if I select a sub-cube in SoFiA, but make the size of that sub-cube equal to that of the full-sized cube, then SoFiA requires “only” about 2 times the cube size in memory to read in the cube rather than 3 times! The overall peak memory usage is just under 4 times the cube size, which fits comfortably into the 60 GB memory allocation.

In other words, if we read a full cube rather than a sub-cube, then SoFiA (or Astropy?) creates an additional full-sized copy of the entire cube in memory! That’s is what pushed SoFiA over the edge in my previous tests. I will now have a look at both the source code as well as Astropy to investigate how this insane behaviour comes about. In the meantime, the workaround is to read a sub-cube of the same size as the full cube to avoid running into memory issues.

**Comment by SoFiA-Admin** [comment]
Okay, I solved the problem. The issue is with SoFiA’s `import_data.py`, more specifically with the lines:

```python
np_Cube = dict_Header['BSCALE'] * np_Cube
np_Cube = dict_Header['BZERO'] + np_Cube
```

That’s a pretty crazy thing to do, as it created two copies of the data cube in memory! I checked this with a 2 GB test cube, and indeed this code occupies 6 GB of memory! Curiously, when reading a sub-cube of the same size of the full cube, it only needs 4 GB, which explains why we have this discrepancy in my earlier tests on the 12.6 GB cube. No idea why reading a sub-cube makes a difference at all...

When I change the line to:

```python
np_Cube *= dict_Header['BSCALE']
np_Cube += dict_Header['BZERO']
```

then no additional memory is used, as no copy is ever made, and the memory requirement shrinks to the expected 2 GB, equal to the size of the cube.

A related question is why we are manually correcting for `BSCALE` and `BZERO` to begin with, as Astropy’s `fits.open()` would scale the data automatically if we hadn’t explicitly disabled this with `do_not_scale_image_data=True`. Does anyone know why this was done?

**Comment by paoloserra** [comment]
> A related question is why we are manually correcting for BSCALE and BZERO to begin with, as Astropy’s fits.open() would scale the data automatically if we hadn’t explicitly disabled this with do_not_scale_image_data=True. Does anyone know why this was done?

See https://github.com/SoFiA-Admin/SoFiA/issues/80

**Comment by RussellJurek** [comment]
I have a vague recollection that it was another memory spike issue. 

Cheers,
Russell

**Comment by ngiese** [comment]
```
np_Cube *= dict_Header['BSCALE']
np_Cube += dict_Header['BZERO']
```
was resulting in an error for users with numpy version 1.11.2 (please see: commit https://github.com/SoFiA-Admin/SoFiA/commit/670377bdc9f8029f305ac177fd4f3091bceffa1c)
I am not happy with the change if it breaks SoFiA for users with numpy versions, that we support. We should find a better solution or at least make sure, that the variables on both sides of the operation are of the same type.

**Comment by paoloserra** [comment]
Issue with scaling fixed by https://github.com/SoFiA-Admin/SoFiA/commit/e8dfa230005476eb1fbda34a9a55e70aef948260 without any need of extra memory. Please test.

**Comment by SoFiA-Admin** [comment]
Thanks everyone for commenting on this issue. In my opinion, the current solution would not appear to make much sense. We now define

```python
np_Cube *= array(dict_Header['BSCALE']).astype(np_Cube.dtype)
np_Cube += array(dict_Header['BZERO']).astype(np_Cube.dtype)
```

If the original FITS file is an integer array (which was the issue originally reported in 670377bdc9f8029f305ac177fd4f3091bceffa1c), then we are effectively turning arbitrary floating point values `BSCALE` and `BZERO` into integers, which would lead to nonsensical results.

I also don’t understand the discussion about the need to manually scale by `BSCALE` and `BZERO` in #80. The discussion there was about memory mapping and the fact that this didn’t work correctly with some versions of Astropy. The decision in the end was to disable memory mapping altogether and simply load the whole cube (or a sub-cube thereof) into memory, so in my opinion we can dump this entire manual scaling of the data and simply let Astropy do that job automatically when we read in the cube with `fits.open()`. After all, we are doing the scaling right after having read in the data, with no additional code (other than closing the input file) in between.

So, my proposal is to simply remove the manual scaling and instead open the file with
```python
fits.open(inFile, memmap=False, do_not_scale_image_data=False)
```
Are there any objections against this approach?

**Comment by SoFiA-Admin** [comment]
I think `memmap` may need to remain as `False`, as I seem to remember that the default might be `True`, at least in some versions of Astropy. The Astropy documentation has the default set to `None`, which makes little sense, as the parameter needs to be a Boolean.

So, I suggest to remove just the `do_not_scale_image_data` option and the two lines of manual scaling, but leave `memmap=False` just to be on the safe side. I will do some testing on my machine to make sure that those changes don’t break anything.

**Comment by SoFiA-Admin** [comment]
Alright, I uploaded the proposed change, so please feel free to test. I ran a few quick tests on my machine, indicating that everything works as expected. In particular, there is no excess memory usage at all with Astropy’s automatic scaling. If everything is confirmed to be working fine, this issue can be marked as resolved.

**Comment by SoFiA-Admin** [comment]
Adding to the discussion above, I managed to also reduce the memory usage of the noise scaling module. The RMS cube produced by that module was originally created with a `dtype` argument. In this case, NumPy defaults to `float64`, which created additional memory overhead for 32-bit floating point data cubes. I changed this such that the RMS cube has the same `dtype` as the original data cube, meaning that the noise scaling module will now require 2 times the cube size in memory irrespective of that the data type of the cube is.

I will close this issue now, as all memory spikes have successfully been removed, and I can now run a 13 GB ASKAP cube on 60 GB of memory without termination of the process.
